[FEATURE] Provide PSR-7 request in Extbase validators (#5434)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ExtensionArchitecture/Extbase/Reference/Domain/Validator.rst

:navigation-title: Validator

..  include:: /Includes.rst.txt
..  index:: Extbase; Validator
..  _extbase_domain_validator:

=========================
Custom Extbase Validators
=========================

..  seealso::

>     *  :ref:`extbase_validation` for general validation in Extbase.

Reviewer: Wrong indent? A list with just one entry?

>     *  :ref:`extbase_validation` for general validation in Extbase.

Author: ```suggestion
    *   :ref:`extbase_validation` for general validation in Extbase.
```

Custom validators are located in the directory :file:`Classes/Domain/Validator`
and therefore in the namespace :php:`MyVendor\MyExtension\Domain\Validator`.

All validators must implement :php-short:`TYPO3\CMS\Extbase\Validation\Validator\ValidatorInterface`.
The usually extend the :php-short:`\TYPO3\CMS\Extbase\Validation\Validator\AbstractValidator`.

..  note::

>     In the package :php:`\TYPO3\CMS\Extbase\Validation\Validator\*` Extbase

Reviewer: "package" is more composer package like typo3/cms-exbase. I would prefer using folder or directory:

```suggestion
    In the directory :php:`\TYPO3\CMS\Extbase\Validation\Validator\*` Extbase
```

..  literalinclude:: _CustomValidator/_PropertyValidatorUsage.php
    :caption: EXT:blog_example/Classes/Domain/Model/Blog.php, modified

..  note::
    Validators added to a property of a model are executed whenever an object
    of that model is passed to a controller action as a parameter.

>     The validation of the parameter can be disabled by the annotation

Author: All I did here was fix the indentation, did not investigate. Would you like to make a follow up?

>     The validation of the parameter can be disabled by the annotation

Reviewer 2: ```suggestion
    The validation result of the parameter can be ignored by using the annotation
```

?

>     The validation of the parameter can be disabled by the annotation

Reviewer 2: see my suggestion?

>     The validation of the parameter can be disabled by the annotation

Reviewer: Thank you. Merged

..  literalinclude:: _CustomValidator/_ObjectValidatorUsage.php
    :caption: EXT:blog_example/Classes/Controller/BlogController.php, modified

..  _extbase_domain_validator-di:

Dependency injection in validators
==================================

Extbase validators are capable of :ref:`dependency injection <Dependency-Injection>`
without further configuration, you can use the constructor method:

..  literalinclude:: _Validator/_MyCustomValidator.php

>     :caption: EXT:my_extension/Classes/Validators/MyCustomValidator.php

Reviewer: To be compatible with your saying somewhere above I prefer adding "Domain" here, too.

```suggestion
    :caption: EXT:my_extension/Classes/Domain/Validators/MyCustomValidator.php
```

..  _extbase_domain_validator-request:

Request object in Extbase validators
====================================

..  versionadded:: 13.2
    Extbase :php-short:`\TYPO3\CMS\Extbase\Validation\Validator\AbstractValidator`
    provides a getter and a setter for the PSR-7 Request object.

You can use the PSR-7 request object in a validator, for example to get
the site settings:

> 

Reviewer: You already have written that at the very top

Author: Not really,  the version added information is in a box and will be removed in  the future. The second sentence is the introduction of the Example and will stay.

..  literalinclude:: _Validator/_RequestValidator.php

>     :caption: EXT:my_extension/Classes/Validators/MyCustomValidator.php

Reviewer: ```suggestion
    :caption: EXT:my_extension/Classes/Domain/Validators/MyCustomValidator.php
```
